### PR TITLE
Created mining pool stats page

### PR DESCRIPTION
**Pull request by nymkappa**
# Description

This PR adds a new page at `/mining/pool/:poolId` and is used to show the data noted below. Note that this is just a PoC/building block and that much more thinking has to be put into what we want to show, how we want to show it, as well as the html/css... (see screenshot for reference)

* Known addresses
* Known coinbase tag
* Mined block count based on the user selected time span
* Empty block count based on the user selected time span
* Mined block list with infinite scroll support

Also, 4 new API endpoints were added:
* `/api/v1/mining/pool/:poolId'` - Get pool stats for all indexed block
* `/api/v1/mining/pool/:poolId/:interval'` - Get pool stats for a specific interval
* `/api/v1/mining/pool-blocks/:poolId'` - Get the last 10 blocks mined by a pool
* `/api/v1/mining/pool-blocks/:poolId/:height'` - Get 10 blocks mined by a pool older than specified block height

image
image

I also did a bit of refactoring in my queries to make sure we don't use unfiltered user input.
Mining related API endpoints are now only available for Bitcoin networks and if the database is enabled.

# Testing

To access the new page, open the mining overview page from the menu, then you can click on individual pools in the list.

This PR does not change the data structure for blocks, nor modify any backend interfaces, so no special action should be required.
Note that the `reward` is currently showing 0 always, as this new field was created in another PR (#1235) and is currently under review.

### CODE REVIEW

**Comment by wiz** [comment]
Cool, maybe we can add the logo too if it exists?

**Comment by wiz** [comment]
```
/api/v1/mining/pool-blocks/:poolId' - Get the last 10 blocks mined by a pool
```

wouldn't it make more sense for that API to follow the others, i.e. `/api/v1/mining/pool/:id/blocks`

File: frontend/src/app/components/pool/pool.component.html [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      <th style="width: 15%;" i18n="latest-blocks.height">Height</th>
+      <th class="d-none d-md-block" style="width: 20%;" i18n="latest-blocks.timestamp">Timestamp</th>
+      <th style="width: 20%;" i18n="latest-blocks.mined">Mined</th>
+      <th style="width: 10%;" i18n="latest-blocks.reward">Reward</th>
+      <th class="d-none d-lg-block" style="width: 15%;" i18n="latest-blocks.transactions">Transactions</th>
+      <th style="width: 20%;" i18n="latest-blocks.size">Size</th>
+    </thead>
+    <tbody *ngIf="blocks$ | async as blocks">
+      <tr *ngFor="let block of blocks">
+        <td><a [routerLink]="['/block' | relativeUrl, block.id]" [state]="{ data: { block: block } }">{{ block.height }}</a></td>
```
Comment (softsimon) [inline frontend/src/app/components/pool/pool.component.html:98]: `[state]=` causes crash on block page due to different block object

File: frontend/src/app/components/pool/pool.component.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import { delay, distinctUntilChanged, map, scan, startWith, switchMap, tap } from 'rxjs/operators';
+import { BlockExtended, PoolStat } from 'src/app/interfaces/node-api.interface';
+import { ApiService } from 'src/app/services/api.service';
+import { StateService } from 'src/app/services/state.service';
+
+@Component({
+  selector: 'app-pool',
+  templateUrl: './pool.component.html',
+  styleUrls: ['./pool.component.scss']
+})
```
Comment (softsimon) [inline frontend/src/app/components/pool/pool.component.ts:15]: See if we can use `changeDetection: ChangeDetectionStrategy.OnPush` here

File: frontend/src/app/components/pool/pool.component.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+export class PoolComponent implements OnInit {
+  poolStats$: Observable<PoolStat>;
+  blocks$: Observable<BlockExtended[]>;
+
+  fromHeight: number = -1;
+  fromHeightSubject: BehaviorSubject<number> = new BehaviorSubject(this.fromHeight);
+
+  blocks: BlockExtended[] = [];
+  poolId: number = undefined;
+  isLoading = false;
```
Comment (softsimon) [inline frontend/src/app/components/pool/pool.component.ts:25]: isLoading not used

File: frontend/src/app/services/api.service.ts [hunk 4]
```diff
@@ -132,4 +132,18 @@ export class ApiService {
...
   }
+
+  getPoolStats$(poolId: number, interval: string | null): Observable<PoolStat> {
+    return this.httpClient.get<PoolStat>(this.apiBaseUrl + this.apiBasePath + `/api/v1/mining/pool/${poolId}/${interval}`);
+  }
+
+  getPoolBlocks$(poolId: number, fromHeight: number): Observable<BlockExtended[]> {
+    if (fromHeight !== undefined) {
+      return this.httpClient.get<BlockExtended[]>(this.apiBaseUrl + this.apiBasePath +
+        `/api/v1/mining/pool/${poolId}/blocks/${fromHeight}`);
```
Comment (softsimon) [inline frontend/src/app/services/api.service.ts:143]: Just use something like `fromHeight ? '/' + fromHeight : ''`

**Review by softsimon — CHANGES_REQUESTED** [review]
Left some comments

File: frontend/src/app/components/pool/pool.component.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        switchMap((params: any) => {
+          this.poolId = params[0];
+          if (this.blocks.length === 0) {
+            this.fromHeightSubject.next(undefined);
+          }
+          return this.apiService.getPoolStats$(this.poolId, params[1] ?? '1w');
+        }),
+        map((poolStats) => {
+          let regexes = '"';
+          for (const regex of JSON.parse(poolStats.pool.regexes)) {
```
Comment (softsimon) [inline frontend/src/app/components/pool/pool.component.ts:53]: Move the JSON.parse( to the backend?

File: frontend/src/app/components/pool/pool.component.html (hunk 1 above)
Comment (nymkappa) [reply frontend/src/app/components/pool/pool.component.html]: Fixed in https://github.com/mempool/mempool/pull/1242/commits/1e96c93557dafe2eef5460ceb23e5c1b09f3e741

File: frontend/src/app/components/pool/pool.component.ts (hunk 3 above)
Comment (nymkappa) [reply frontend/src/app/components/pool/pool.component.ts]: Fixed in Fixed in https://github.com/mempool/mempool/pull/1242/commits/1e96c93557dafe2eef5460ceb23e5c1b09f3e741

File: frontend/src/app/services/api.service.ts (hunk 4 above)
Comment (nymkappa) [reply frontend/src/app/services/api.service.ts]: Fixed in Fixed in https://github.com/mempool/mempool/pull/1242/commits/1e96c93557dafe2eef5460ceb23e5c1b09f3e741

File: frontend/src/app/components/pool/pool.component.ts (hunk 5 above)
Comment (nymkappa) [reply frontend/src/app/components/pool/pool.component.ts]: Fixed in https://github.com/mempool/mempool/pull/1242/commits/fa8607c57d30880abcb17b8a5ed69db462da7012

File: frontend/src/app/components/pool/pool.component.ts (hunk 2 above)
Comment (nymkappa) [reply frontend/src/app/components/pool/pool.component.ts:15]: Fixed in https://github.com/mempool/mempool/pull/1242/commits/1e96c93557dafe2eef5460ceb23e5c1b09f3e741

**Review by softsimon — APPROVED** [review]
tested ACK

**Pull request merged**